rust/abacus-base/src/metrics/core.rs

```diff
@@ -292,65 +263,117 @@ impl CoreMetrics {
         )?)
     }
 
-    /// Call with RPC duration after it is complete
-    pub fn rpc_complete(&self, chain: &str, method: &str, duration: Duration) {
-        self.rpc_latencies
-            .with_label_values(&[chain, method, &self.agent_name])
-            .observe(duration.as_secs_f64())
-    }
-
-    /// Gauge for measuring the last known message leaf index
+    /// Reports the current highest message leaf index at multiple phases of the
+    /// relaying process. There may be messages that have not reached a certain
+    /// stage, such as being fully processed, even if the reported index is
+    /// higher than that message's index.
+    ///
+    /// Some phases are not able to report the remote chain, but origin chain is
+    /// always reported.
+    ///
+    /// Labels:
+    /// - `phase`: The the leaf index is being tracked at, see below.
+    /// - `origin`: Origin chain the index comes from.
+    /// - `remote`: Remote chain for the message. This will skip values because
+    ///   the indexes are contiguous by origin not remote.
+    ///
+    /// The following phases are implemented:
+    /// - `dispatch`: Highest leaf index which has been indexed on the outbox
+    ///   contract syncer and stored in the relayer DB.
+    /// - `signed_offchain_checkpoint`: Highest leaf index of a checkpoint which
+    ///   is known to have been signed by a quorum of validators.
+    /// - `processor_loop`: Highest leaf index which the MessageProcessor loop
+    ///   has gotten to but not attempted to send it.
+    /// - `message_processed`: When a leaf index was processed as part of the
+    ///   MessageProcessor loop.
     pub fn last_known_message_leaf_index(&self) -> IntGaugeVec {
         self.last_known_message_leaf_index.clone()
     }
 
-    /// Gauge for reporting the current outbox state.
+    /// Gauge for reporting the current outbox state. This is either 0 (for
+    /// UnInitialized), 1 (for Active), or 2 (for Failed). These are from the
+    /// outbox contract States enum.
+    ///
+    /// Labels:
+    /// - `chain`: The chain the outbox is for.
     pub fn outbox_state(&self) -> IntGaugeVec {
         self.outbox_state.clone()
     }
 
-    /// Gauge for the latest checkpoint at various phases.
+    /// Latest message leaf index observed on a checkpoint on the
```

Comment:
Only the first phase is for observation on the outbox, the other phase isn't related to the contract no?

Reply:
Correct, updating.